ext/http/01_http.js

```diff
@@ -74,104 +75,82 @@ class HttpConn {
-      this.close();
-      // A connection error seen here would cause disrupted responses to throw
-      // a generic `BadResource` error. Instead store this error and replace
-      // those with it.
-      this[connErrorSymbol] = error;
-      if (
-        ObjectPrototypeIsPrototypeOf(BadResourcePrototype, error) ||
-        ObjectPrototypeIsPrototypeOf(InterruptedPrototype, error) ||
-        StringPrototypeIncludes(error.message, "connection closed")
-      ) {
+    while (true) {
+      if (this.#error) {
+        throw this.#error;
+      }
+      const conn = this.#connections.pop();
+      if (conn) {
+        return { request: conn[1], respondWith: (resp) => { 
+          conn[0](resp);
+          // TODO(mmastrac): How do we resolve this promise? Do we?
+          return new Promise(r => r());
+        } };
+      }
+      const wakeup = new Promise(r => this.#waiters.push(r));
+      if (await wakeup === false) {
         return null;
       }
-      throw error;
-    }
-    if (nextRequest == null) {
-      // Work-around for servers (deno_std/http in particular) that call
-      // `nextRequest()` before upgrading a previous request which has a
-      // `connection: upgrade` header.
-      await null;
-
-      this.close();
-      return null;
     }
-
-    const { 0: streamRid, 1: method, 2: url } = nextRequest;
-    SetPrototypeAdd(this.managedResources, streamRid);
-
-    /** @type {ReadableStream<Uint8Array> | undefined} */
-    let body = null;
-    // There might be a body, but we don't expose it for GET/HEAD requests.
-    // It will be closed automatically once the request has been handled and
-    // the response has been sent.
-    if (method !== "GET" && method !== "HEAD") {
-      body = readableStreamForRid(streamRid, false);
-    }
-
-    const innerRequest = newInnerRequest(
-      method,
-      url,
-      () => ops.op_http_headers(streamRid),
-      body !== null ? new InnerBody(body) : null,
-      false,
-    );
-    innerRequest[streamRid] = streamRid;
-    const abortController = new AbortController();
-    const request = fromInnerRequest(
-      innerRequest,
-      abortController.signal,
-      "immutable",
-      false,
-    );
-
-    const respondWith = createRespondWith(
-      this,
-      streamRid,
-      request,
-      this.#remoteAddr,
-      this.#localAddr,
-      abortController,
-    );
-
-    return { request, respondWith };
   }
 
   /** @returns {void} */
   close() {
-    if (!this.#closed) {
-      this.#closed = true;
-      core.close(this.#rid);
-      for (const rid of new SafeSetIterator(this.managedResources)) {
-        SetPrototypeDelete(this.managedResources, rid);
-        core.close(rid);
-      }
+    for (const waiter of this.#waiters) {
+      waiter(false);
+    }
+    for (const request of this.#requests) {
+      request.close();
     }
+    this.#requests = new Set();
+    // Don't allow further listening
+    this.#onError(new TypeError("closed"));
```

Comment:
This seems funky - why not use a boolean flat like `this.#closed` instead of throwing a specific error? What if user code throws `new TypeError("closed")`?

Reply:
I think `#closed` is a better approach.. I'll add this to the list of things to fix.